### PR TITLE
Add testutils for accessing all storage of a contract

### LINKED ISSUE #737: Test utils to fetch all keys in a contract's data
### What problem does your feature solve?
Having a testutil in the SDK to retrieve all keys in a contract's data could ease testing.

(context: https://discord.com/channels/897514728459468821/1033024797153706085/1033055285507543141)

### PR DESCRIPTION
**Pull request by leighmcculloch**
### What
Add testutils for accessing all storage of a contract.

### Why
Sometimes when testing it is convenient to assert on all storage of a contract.

Close #737

### CODE REVIEW

File: soroban-sdk/src/storage.rs [hunk 1]
```diff
@@ -383,3 +458,49 @@ impl Instance {
...
+                    }),
+                ..
+            } = *v
+            else {
+                continue;
+            };
+            return match storage {
+                Some(map) => {
+                    let map: Val =
+                        Val::try_from_val(env, &xdr::ScVal::Map(Some(map.clone()))).unwrap();
```
Comment (leighmcculloch) [inline soroban-sdk/src/storage.rs:498]: I really didn't want to introduce an ScMap clone in here, but it is too inconvenient to work around. I need it in an owned form for the ScVal unfortunately.

File: soroban-sdk/src/tests/storage_testutils.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        e.storage().persistent().set(&20, &30);
+        e.storage().temporary().set(&100, &200);
+        e.storage().temporary().set(&100, &());
+        e.storage().temporary().set(&200, &300);
+    });
+
+    e.as_contract(&id, || {
+        assert_eq!(
+            e.storage().instance().all(),
+            Map::<Val, Val>::from_array(&e, [(1.into(), true.into()), (2.into(), 3.into())])
```
Comment (leighmcculloch) [inline soroban-sdk/src/tests/storage_testutils.rs:31]: Note that these `.all()` functions, like any storage function, must be called in the context of a contract.

File: soroban-sdk/src/storage.rs [hunk 3]
```diff
@@ -291,6 +294,42 @@ impl Persistent {
     }
 }
 
+#[cfg(any(test, feature = "testutils"))]
+#[cfg_attr(feature = "docs", doc(cfg(feature = "testutils")))]
+impl testutils::storage::Persistent for Persistent {
```
Comment (dmkozh) [inline soroban-sdk/src/storage.rs:299]: Can we create a helper to de-duplicate persistent and temp code? IIUC the only difference is in `durability` filter.

Comment (leighmcculloch) [reply soroban-sdk/src/storage.rs:299]: Yup, although value seems mediocre and may result in harder to maintain code down the line.

Comment (dmkozh) [reply soroban-sdk/src/storage.rs:299]: I don't expect persistent and temp storage interfaces to ever diverge, but up to you really.

**Review by dmkozh — APPROVED** [review]

File: soroban-sdk/src/storage.rs (hunk 3 above)
Comment (leighmcculloch) [reply soroban-sdk/src/storage.rs]: 380eae4

Comment (leighmcculloch) [reply soroban-sdk/src/storage.rs]: That worked well, thanks 👏🏻.

**Pull request merged**